### ISSUE TITLE
Add a way to initializate debug extension before every new debug session

### ISSUE THREAD

**Issue by TylerLeonhardt** [microsoft/vscode#81806]
Version: 1.39.0-insider
Commit: b0d714d2de2886060e0904bf0cc774e0d70667cd
Date: 2019-09-26T09:05:41.420Z
Electron: 6.0.9
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Darwin x64 18.7.0

Steps to Reproduce:

1. open a powershell file with the PowerShell e x tension
2. drop a breakpoint and F5
3. Hit the restart button when the breakpoint is hit

You'll notice nothing happens... this is because `resolveDebugConfiguration` is not run for when the debug adapter is starting up again so it can't set up needed state on the client side.

The 'restart' command isn't supported so this is behavior should be shutting down and starting up the debug adapter again.

Is this by design that it's not running `resolveDebugConfiguration` on restart? If so, any recs on how we can set up state in the extension host side before the debug adapter is launched?

**Comment by isidorn** [comment]
@TylerLeonhardt yes this is by design.
VSCode will only re-read the configuration and resolve it again if the underlying `launch.json` file changed. [Code pointer](https://github.com/Microsoft/vscode/blob/6ff7bf27c9dc447cd95693debdc5b90910e804b5/src/vs/workbench/contrib/debug/browser/debugService.ts#L583)

So as a workaround your extension can just make some edit to the underlying `launch.json`.
However setting up state in the `resolveDebugConfiguration` should not be done. The `resolveDebugConfiguration` is only for resolving the configuration, your extension should optimally not do too much things there.
For example your extension can listen to `onDidStartDebugSession`. Though this might be a bit too late.
If you need to setup your state only once you can be actived `onDebugType` events. However we might be missing a mechanism that you intiialize before every new session. For that you can file a new feature request. Though @weinand might have some ideas which I am missing

**Comment by TylerLeonhardt** [comment]
Does the `onDidStartDebugSession` happen _after_ the debug server and client exchange something? If so, then that's sadly too late.

>  intiialize before every new session

yeah this is exactly what I'm looking for.

**Comment by TylerLeonhardt** [comment]
We can change the title of this issue to maybe "Add a mechanism to allow for initialization before every new debug session"?

**Comment by isidorn** [comment]
Yes, I have renamed it. And it is a fair feature request
fyi @weinand

**Comment by TylerLeonhardt** [comment]
I find it unlikely that this will gain enough traction considering it's only an ask from the extension development side.

I'd would love to see more discussion from the VS Code team side on this as it would still be nice to have available for extension authors.

**Comment by TylerLeonhardt** [comment]
Pls no

**Comment by roblourens** [comment]
Sorry I don't understand the use-case, why do you need this?

If there is some setup to run when the debug adapter starts, then wouldn't the debug adapter code itself do that?

**Comment by TylerLeonhardt** [comment]
Basically there is code inside of `resolveDebugConfiguration` that was needed in order to properly spin up the Debug Adapter... but that code wasn't run on restart so the PowerShell debug session wouldn't work.

**Comment by roblourens** [comment]
That should be in a `DebugAdapterDescriptorFactory` though, `resolveDebugConfiguration` seems the wrong place for it

**Comment by TylerLeonhardt** [comment]
Yeah I tried the powershell extension today and it seems to be working so I think the scenario is covered.
